client.go

```diff
@@ -1375,3 +1397,545 @@ func (c *ToznySDKV3) UnbrokerShare(ctx context.Context, authorizerClientID strin
+		memberReq = append(memberReq, storageClient.GroupMember{
+			ClientID:        uuid.MustParse(c.StorageClient.ClientID),
+			MembershipKey:   membershipKeyResp,
+			CapabilityNames: groupMemberCapabilities,
+		})
+		// add group members
+		addMemberReq := storageClient.AddGroupMembersRequest{
+			GroupID:      createGroupResponse.GroupID,
+			GroupMembers: memberReq,
+		}
+		_, err = c.StorageClient.AddGroupMembers(ctx, addMemberReq)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// the group already exists
+		group = &responseList.Groups[0]
+	}
+	return group, nil
+}
+
+// ValidateSecret checks that the secret contains valid input for each entry
+func ValidateSecret(secret CreateSecretOptions) error {
+	secret.SecretName = strings.TrimSpace(secret.SecretName)
+	secret.SecretValue = strings.TrimSpace(secret.SecretValue)
+	if secret.SecretType == "" {
+		return errors.New("type cannot be empty")
+	}
+	if !SliceContainsString(SECRET_TYPES, secret.SecretType) {
+		return errors.New("invalid type")
+	}
+	if secret.SecretName == "" {
+		return errors.New("name cannot be empty")
+	}
+	matched, err := regexp.MatchString(`^[a-zA-Z0-9-_]{1,50}$`, secret.SecretName)
+	if err != nil {
+		return err
+	}
+	if !matched {
+		return errors.New("Secret name must contain 1-50 alphanumeric characters, -, or _")
+	}
+	if secret.SecretValue == "" && secret.SecretType == "Credential" {
+		return errors.New("Value cannot be empty")
+	}
+	if secret.SecretType == "File" && strings.TrimSpace(secret.SecretName) == "" {
+		return errors.New("File name cannot be empty")
+	}
+	if secret.SecretType == "Client" {
+		err := VerifyRawClientCredentials(secret.SecretValue)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// SliceContainsString checks if str is present in an array of strings
+func SliceContainsString(list []string, str string) bool {
+	for _, v := range list {
+		if v == str {
+			return true
+		}
+	}
+	return false
+}
+
+type ClientConfigRoot struct {
+	APIBaseURL        string      `json:"api_url"`
+	APIKeyID          string      `json:"api_key_id"`
+	APISecret         string      `json:"api_secret"`
+	ClientID          string      `json:"client_id"`
+	ClientEmail       string      `json:"client_email"`
+	PublicKey         string      `json:"public_key"`
+	PrivateKey        string      `json:"private_key"`
+	PublicSigningKey  string      `json:"public_signing_key"`
+	PrivateSigningKey string      `json:"private_signing_key"`
+	Version           interface{} `json:"version"`
+}
+
+// VerifyRawClientCredentials ensures that client credentials are in the proper form and contain all required keys
+func VerifyRawClientCredentials(credential string) error {
+	var clientConfig ClientConfigRoot
+	err := json.Unmarshal([]byte(credential), &clientConfig)
+	if err != nil {
+		return err
+	}
+	values := reflect.ValueOf(clientConfig)
+	keys := values.Type()
+	for index := 0; index < values.NumField(); index++ {
```

Comment:
Which fields does this loop cover that are not covered below? It is probably clearer to explicitly check them

Reply:
it's checking every field, including APIURL, APISecret, and APIKeyID which aren't checked anywhere else.

Reply:
While I think that reflection is clever and generic, in this case would argue that a validate should just do a bunch if checks on known fields.